In [1]:
import numpy as np
import pandas as pd
import seaborn as sns
import matplotlib.pyplot as plt
import warnings
warnings.filterwarnings('ignore')

In [2]:
from sklearn import datasets
diabetes = datasets.load_diabetes()

In [3]:
x = diabetes.data
x=pd.DataFrame(x)
x.shape

(442, 10)

In [4]:
x.head()

0         1         2         3         4         5         6  \
0  0.038076  0.050680  0.061696  0.021872 -0.044223 -0.034821 -0.043401   
1 -0.001882 -0.044642 -0.051474 -0.026328 -0.008449 -0.019163  0.074412   
2  0.085299  0.050680  0.044451 -0.005671 -0.045599 -0.034194 -0.032356   
3 -0.089063 -0.044642 -0.011595 -0.036656  0.012191  0.024991 -0.036038   
4  0.005383 -0.044642 -0.036385  0.021872  0.003935  0.015596  0.008142   

          7         8         9  
0 -0.002592  0.019908 -0.017646  
1 -0.039493 -0.068330 -0.092204  
2 -0.002592  0.002864 -0.025930  
3  0.034309  0.022692 -0.009362  
4 -0.002592 -0.031991 -0.046641

In [5]:
y = diabetes.target
y=pd.DataFrame(y)
y.shape

(442, 1)

In [6]:
y.head()

0
0  151.0
1   75.0
2  141.0
3  206.0
4  135.0

In [7]:
from sklearn.model_selection import train_test_split,cross_val_score,GridSearchCV

In [8]:
x_train,x_test,y_train,y_test=train_test_split(x,y,test_size=0.2,random_state=68)

In [9]:
x_train.shape,x_test.shape

((353, 10), (89, 10))

In [10]:
y_train.shape,y_test.shape

((353, 1), (89, 1))

**RandomForestRegressor (Hyperparameter Tuning)**

In [11]:
from sklearn.ensemble import RandomForestRegressor
RFR=RandomForestRegressor()
RFR.fit(x_train,y_train)

RandomForestRegressor(bootstrap=True, criterion='mse', max_depth=None,
           max_features='auto', max_leaf_nodes=None,
           min_impurity_decrease=0.0, min_impurity_split=None,
           min_samples_leaf=1, min_samples_split=2,
           min_weight_fraction_leaf=0.0, n_estimators=10, n_jobs=None,
           oob_score=False, random_state=None, verbose=0, warm_start=False)

In [12]:
RFR.score(x_test,y_test)

0.44420307962372385

In [13]:
print(cross_val_score(RFR,x,y,cv=5).mean())

0.36752730933903177


In [14]:
parameterRFR = { 
            "n_estimators"      : [10,20,30,40,50,100,200],
            "max_features"      : ["auto", "sqrt", "log2"],
            "max_depth"         : [2,3,4,5,None],
            "min_samples_split" : [2,3,4],
            "min_samples_leaf"  : [ 2, 3,4],
            "bootstrap"         : [True, False],
            }

In [15]:
gridRFR = GridSearchCV(RFR, param_grid=parameterRFR, n_jobs=-1, cv=10)
gridRFR.fit(x_train, y_train)

C:\ProgramData\Anaconda3\lib\site-packages\sklearn\model_selection\_search.py:841: DeprecationWarning: The default of the `iid` parameter will change from True to False in version 0.22 and will be removed in 0.24. This will change numeric results when test-set sizes are unequal.
  DeprecationWarning)


GridSearchCV(cv=10, error_score='raise-deprecating',
       estimator=RandomForestRegressor(bootstrap=True, criterion='mse', max_depth=None,
           max_features='auto', max_leaf_nodes=None,
           min_impurity_decrease=0.0, min_impurity_split=None,
           min_samples_leaf=1, min_samples_split=2,
           min_weight_fraction_leaf=0.0, n_estimators=10, n_jobs=None,
           oob_score=False, random_state=None, verbose=0, warm_start=False),
       fit_params=None, iid='warn', n_jobs=-1,
       param_grid={'n_estimators': [10, 20, 30, 40, 50, 100, 200], 'max_features': ['auto', 'sqrt', 'log2'], 'max_depth': [2, 3, 4, 5, None], 'min_samples_split': [2, 3, 4], 'min_samples_leaf': [2, 3, 4], 'bootstrap': [True, False]},
       pre_dispatch='2*n_jobs', refit=True, return_train_score='warn',
       scoring=None, verbose=0)

In [16]:
gridRFR.best_score_

0.4433900169709411

In [17]:
gridRFR.best_params_

{'bootstrap': True,
 'max_depth': 5,
 'max_features': 'log2',
 'min_samples_leaf': 2,
 'min_samples_split': 3,
 'n_estimators': 10}

In [18]:
RFR=RandomForestRegressor(bootstrap = True,
max_depth = 5,                        
 max_features= 'log2',
 min_samples_leaf = 2,
 min_samples_split= 3,
 n_estimators= 10)

In [19]:
print(cross_val_score(RFR,x,y,cv=5).mean())

0.4157996772430673


**After doing hyperparameter tuning for RandomForestRegressor the cross val score is increase from 0.36 to 0.41**

**GradientBoostingClassifier(Hyperparameter Tuning)**

In [20]:
from sklearn.ensemble import GradientBoostingRegressor
GBR=GradientBoostingRegressor()
GBR.fit(x_train,y_train)

GradientBoostingRegressor(alpha=0.9, criterion='friedman_mse', init=None,
             learning_rate=0.1, loss='ls', max_depth=3, max_features=None,
             max_leaf_nodes=None, min_impurity_decrease=0.0,
             min_impurity_split=None, min_samples_leaf=1,
             min_samples_split=2, min_weight_fraction_leaf=0.0,
             n_estimators=100, n_iter_no_change=None, presort='auto',
             random_state=None, subsample=1.0, tol=0.0001,
             validation_fraction=0.1, verbose=0, warm_start=False)

In [21]:
GBR.score(x_test,y_test)

0.4256536582284155

In [22]:
print(cross_val_score(GBR,x,y,cv=5).mean())

0.4058088674013577


In [23]:
parameterGBR={'n_estimators':[20,50,100,200,500], 
            'learning_rate': [0.1,0.05,0.02],
            'max_depth':[2,3,4,5,6], 
            'min_samples_leaf':[3,2,1],
              'alpha':[0.5,0.9,0.8,0.6]
            }

In [24]:
gridGBR = GridSearchCV(GBR, param_grid=parameterGBR, n_jobs=-1, cv=10)
gridGBR.fit(x_train, y_train)

C:\ProgramData\Anaconda3\lib\site-packages\sklearn\model_selection\_search.py:841: DeprecationWarning: The default of the `iid` parameter will change from True to False in version 0.22 and will be removed in 0.24. This will change numeric results when test-set sizes are unequal.
  DeprecationWarning)


GridSearchCV(cv=10, error_score='raise-deprecating',
       estimator=GradientBoostingRegressor(alpha=0.9, criterion='friedman_mse', init=None,
             learning_rate=0.1, loss='ls', max_depth=3, max_features=None,
             max_leaf_nodes=None, min_impurity_decrease=0.0,
             min_impurity_split=None, min_samples_leaf=1,
             min_sampl...=None, subsample=1.0, tol=0.0001,
             validation_fraction=0.1, verbose=0, warm_start=False),
       fit_params=None, iid='warn', n_jobs=-1,
       param_grid={'n_estimators': [20, 50, 100, 200, 500], 'learning_rate': [0.1, 0.05, 0.02], 'max_depth': [2, 3, 4, 5, 6], 'min_samples_leaf': [3, 2, 1], 'alpha': [0.5, 0.9, 0.8, 0.6]},
       pre_dispatch='2*n_jobs', refit=True, return_train_score='warn',
       scoring=None, verbose=0)

In [25]:
gridGBR.best_score_ , gridGBR.best_params_

(0.44282037004221814,
 {'alpha': 0.9,
  'learning_rate': 0.1,
  'max_depth': 2,
  'min_samples_leaf': 1,
  'n_estimators': 50})

In [26]:
GBR=GradientBoostingRegressor(learning_rate= 0.1,
  max_depth= 2,
  max_features= 1.0,
  min_samples_leaf= 1,
  n_estimators= 50,
  alpha=0.8)

In [27]:
print(cross_val_score(GBR,x,y,cv=5).mean())

0.4397119301698561


**After doing hyperparameter tuning for GradientBoostingRegressor the cross val score is increase from 0.40 to 0.43**